### PR TITLE
Claiming Rewards from the Reward Pot

### LINKED ISSUE #95: Claiming Rewards from the Reward Pot
### 10.2 Claiming Rewards from the Reward Pot
Rewards accumulate in the rewards pot. To trigger an actual payout to users (i.e. to make rewards claimable) a special type of proposal is made, proposing that all users should receive a payout based on the reward pot’s holdings.
This reward payout proposal includes the specific currency that should be paid, and only one currency is handled at a time. In the event that the proposal is approved by vote of reputation, then all user’s tokens are locked until they claim their payout. Locking is necessary, because the token balance of each account factors into the rewards formula of Equation (7). Locking is triggered by incrementing the colony’s ‘most recent payout’ counter.
Our currency contract contains a locking mechanism ensuring that a user cannot move tokens while they have (token-weighted) votes to reveal; we use the same mechanism here to ensure that a user cannot move tokens after a payout is approved by the members of the colony but before the user has claimed their rewards. The colony has a counter for each user that is incremented whenever they claim a payout; they can also waive their claim to a payout that will increment this counter.
While it is of course up to the members of each individual colony to decide, it is advisable that these payout proposals should only be accepted sporadically to keep the gas costs low for the users claiming their payouts, as well as simply to not be a nuisance to the users continually finding their tokens locked.
**Rewards are only available to accounts that hold both tokens and reputation**, and the amount claimable by each account depends on both token balance and reputation (see equation (7) below). Therefore we need to have a similar behaviour to ‘lock’ the reputation of the users for the payout. When a payout is activated, the current state of the reputation tree is recorded in the payout itself. Users are paid out according to their reputation in this state, rather than the most recent state, to ensure all users get an appropriate payout and to avoid exploiting the system (which might otherwise be possible via e.g. delaying reward collection until after completing a task, increasing their reputation).

**The Rewards Formula**
The amount that each user (ui) of a colony (C) is entitled to claim (pi) is a function of their colony
token holdings (ti) and their total reputation in the colony (ri):
screen shot 2017-10-03 at 13 21 14
This is a (normalised) geometric average of the user’s token holdings and reputation. We note that this is very unlikely to payout all the tokens set aside for a payout — the only way it would do so is if everyone had the same proportion of reputation in the colony as they did proportion of tokens in the colony. However, the geometric average is the natural way to fairly capture the influence of two variables with different ranges, and ensures that large token holders must earn large amounts of reputation to get the most from the payouts. The total reputation and user reputation in the colony are all provable on-chain at claim time via a Merkle proof that the `ReputationRootHash` (Section 7) contains some values claimed by the user; the user’s balance of colony tokens and the total number of tokens issued is trivial to lookup.
After some sufficiently long period of time (500000 blocks), all unclaimed tokens can be reclaimed on behalf of the colony by a user, and the payout closed. Any users that have not claimed their payout by that point will still have their tokens locked, and they will remain locked until they issue a transaction waiving their claim to the payout (indeed, they already passively did this by not claiming it in a timely fashion). Unclaimed tokens are returned to the reward pot and become part of the next reward cycle.

### PR DESCRIPTION
**Pull request by filiplazovic**
Closes #95 

# Reward payout

Storage variables:
- Global reward payout counter - `globalRewardPayoutCount`
- Active reward payouts - `activeRewardPayouts`
- User reward payout counter mapping - `userRewardPayoutCount`
- Reward payout cycles mapping - `rewardPayoutCycles`

Implemented functions for:
- starting the payout cycle - `startNextRewardPayout`
- claiming reward payout - `claimRewardPayout`
- waiving reward - `waiveRewardPayout`
- moving remaining funds after payout - `moveRemainingRewardPayoutFunds`
- getting the info about payout - `getRewardPayoutInfo`

## Starting the payout cycle

Takes one argument: token address

Saving all the necessary data to `rewardPayoutCycles` (current root hash, current total tokens, reward pot balance, token, current block number).
Incrementing `globalRewardPayoutCount` and using it for mapping key.
Setting payout token to active, which will not allow creation of new payouts with the same token until the current one finishes.

## Claiming the reward payout

Takes three arguments: payout id, user reputation, colony-wide reputation

Can be called only if reward payout is active i.e if its created less than 60 days ago.
Can be called only if payout is not claimed, i.e if `payoutId - userRewardPayoutCount = 1` (To force the user to claim the payout that is next in queue, otherwise there would be a lot of problems).

Calculating amount to be send to user according to formula defined in the whitepaper.
Formula
Incrementing `userRewardPayoutCount`.
Sending the user the calculated amount.

## Waiving reward payout

Takes one argument: number of payouts to waive

Can only be called if user has not claimed that reward i.e if `payoutId - userRewardPayoutCount = 1`.

Incrementing `userRewardPayoutCount`.

## Moving remaining funds after payout

Can be called only if payout is finished i.e if its created more than 60 days ago.
Setting payout token to inactive. This will allow creation of reward payout with that token.

Moves remaining funds to pot 0.

## Getting the info about the payout
Takes one argument: payout id

Getting the useful information about the payout for UI.

### CODE REVIEW

File: contracts/ColonyFunding.sol [hunk 1]
```diff
@@ -141,6 +142,86 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+
+    //TODO: Unlock user tokens
+
+    rewardPayoutCycles[_payoutId].remaining = sub(rewardPayoutCycles[_payoutId].remaining, reward);
+
+    payoutToken.transfer(msg.sender, reward);
+  }
+
+  function waiveRewardPayout(uint8 _payoutId) public {
+    require(_payoutId - userRewardPayoutCount[msg.sender] == 1);
```
Comment (area) [inline contracts/ColonyFunding.sol]: Arguably, it should be possible for a user to waive multiple payouts at once - if they've been away for a long time, maybe there are lots for them to waive.

File: contracts/ColonyStorage.sol [hunk 2]
```diff
@@ -45,6 +45,18 @@ contract ColonyStorage is DSAuth {
   // Pot 0 is the 'reward' pot containing funds that can be paid to holders of colony tokens in the future.
   mapping (uint256 => Pot) pots;
 
+  struct RewardPayoutCycle {
```
Comment (area) [inline contracts/ColonyStorage.sol]: Document the properties of this struct.

File: contracts/ColonyFunding.sol [hunk 3]
```diff
@@ -141,6 +142,86 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+  function claimRewardPayout(uint8 _payoutId, int256 _userReputation, int256 _totalReputation) public {
+    require(block.number - rewardPayoutCycles[_payoutId].blockNumber < 500000);
+    require(_payoutId - userRewardPayoutCount[msg.sender] == 1);
+
+    //TODO: Prove that userReputation and totalReputation in reputationState are correct
+
+    address tokenAddress = rewardPayoutCycles[_payoutId].token;
+    ERC20Extended payoutToken = ERC20Extended(tokenAddress);
+    uint256 userTokens = payoutToken.balanceOf(msg.sender);
+    uint256 totalTokens = sub(payoutToken.totalSupply(), payoutToken.balanceOf(address(this)));
```
Comment (area) [inline contracts/ColonyFunding.sol]: 1. The tokens used to work out the reward a user is entitled to should be the colony tokens they hold in the colony paying out the reward, not the tokens of the type being paid out that they already hold.

2. I hadn't considered removing the colony tokens held by the colony from consideration from this calculation. I think that this will give strictly better behaviour than what was spec'd (everyone should get slightly more rewards than they would have done before - i.e. there will be less left over and returned to the pot). Very nice!

**Comment by area** [comment]
The approximations with the payout function is to be expected - even if the calculation could be done precisely on-chain, the payouts could not. I suspect what you've done is accurate enough, but maybe test with very small numbers for both the tokens and the reputation is in order just to see how bad it can be.

Parallel payouts with one token is an interesting one. I like that you've taken that possibility in to account so far; given that an additional mapping would be required, and we want to encourage payouts to be fairly infrequent, I think I'm okay with disallowing it. I will check with some others and confirm.

**Comment by filiplazovic** [comment]
Playing with ganache was useful, although the code was rusty.

Thanks for the feedback. I'm just not sure what you meant by:
> even if the calculation could be done precisely on-chain, the payouts could not

**Comment by area** [comment]
>Thanks for the feedback. I'm just not sure what you meant by:
>>    even if the calculation could be done precisely on-chain, the payouts could not

Just that there is a smallest indivisible token unit on-chain, so even if the calculation could be done precisely and required, for example, .1 tokens to be paid out, we couldn't do that.

File: contracts/IColony.sol [hunk 4]
```diff
@@ -276,7 +276,48 @@ contract IColony {
...
-  
+
+  /// @notice Start next reward payout in `_token`. All funds in that are in reward pot at the time will become unavailable.
+  /// All tokens will be locked, and can be unlocked by calling `waiveRewardPayout` or `claimRewardPayout`.
+  /// @param _token Addess of the token used for reward payout
+  function startNextRewardPayout(address _token) public returns (uint256);
+
+  /// @notice Claim the reward payout at `_payoutId`. User needs to provide his reputation and colony-wide reputation
+  /// which will be proven via Merkle proof inside this function.
+  /// Can only be called if payout is active, i.e its created less than 500k blocks ago.
```
Comment (area) [inline contracts/IColony.sol]: This comment should talk about time, not blocks.

File: contracts/IColony.sol [hunk 5]
```diff
@@ -276,7 +276,48 @@ contract IColony {
...
+  /// @param _payoutId Id of the reward payout
+  /// @return Reputation root hash at the time of creation
+  /// @return Total amount of tokens taken aside for reward payout
+  /// @return Remaining (unclaimed) amount of tokens
+  /// @return Token address
+  /// @return Block number at the time of creation
+  function getRewardPayoutInfo(uint8 _payoutId) public view returns (bytes32, uint256, uint256, address, uint256);
+
+  /// @notice Moves remaining (unclaimed) tokens back to reward pot
+  /// Can only be called when reward payout cycle is finished i.e its created more than 500k blocks ago
```
Comment (area) [inline contracts/IColony.sol]: Again, 60 days

File: contracts/ColonyFunding.sol [hunk 6]
```diff
@@ -141,6 +142,89 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+      amount,
+      _token,
+      block.timestamp
+    );
+
+    return globalRewardPayoutCount;
+  }
+
+  function claimRewardPayout(uint8 _payoutId, int256 _userReputation, int256 _totalReputation) public {
+    require(block.timestamp - rewardPayoutCycles[_payoutId].blockTimestamp < 5184000000); // 60 days
```
Comment (area) [inline contracts/ColonyFunding.sol]: `block.timestamp` is in seconds, not milliseconds. So this is actually 60000 days! While technically you won't need to change the tests (their very large numbers will still take us past the reward payout window once this is fixed), I'd prefer it if they were changed too.

File: contracts/ColonyFunding.sol [hunk 7]
```diff
@@ -141,6 +142,89 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+    require(_payoutId - userRewardPayoutCount[msg.sender] == 1);
+
+    //TODO: Prove that userReputation and totalReputation in reputationState are correct
+
+    uint256 userTokens = token.balanceOf(msg.sender);
+    uint256 totalTokens = sub(token.totalSupply(), token.balanceOf(address(this)));
+
+    require(userTokens > 0 && totalTokens > 0 && _userReputation > 0 && _totalReputation > 0);
+
+    uint256 diff = 76 - SafeMath.numDigits(totalTokens * uint256(_totalReputation));
```
Comment (area) [inline contracts/ColonyFunding.sol]: `uint256` can hold up to a 78 digit number in base 10, so this can underflow to a very large value, which I'm not terribly happy with. I *think* the worse case scenario there is that `reward` ends up being zero (i.e. someone gets less than they should, rather than more than they should), but users would be pretty annoyed about that. Did this implementation come from anywhere in particular? Why did you choose 76?

File: test/colony-funding.js [hunk 8]
```diff
@@ -403,4 +407,367 @@ contract("Colony Funding", () => {
...
+        totalReputation: toBN(Math.floor(Math.random() * 1e12) + 1e8),
+        userReputation: toBN(Math.floor(Math.random() * 1e8)),
+        delta: 10
+      },
+      {
+        totalReputation: toBN(Math.floor(Math.random() * 1e21) + 1e16).mul(toBN(Math.floor(Math.random() * 1e3))),
+        userReputation: toBN(Math.floor(Math.random() * 1e16)),
+        delta: 100
+      }
+    ];
```
Comment (area) [inline test/colony-funding.js]: I like these tests! Can we have more of them, for even more extreme values? 😃

File: contracts/ColonyFunding.sol [hunk 9]
```diff
@@ -141,6 +142,89 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+    uint256 factor = (diff % 2 == 0) ? diff : sub(diff, 1);
+    uint256 sharePercentage = SafeMath.sqrt(SafeMath.percent(userTokens * uint256(_userReputation), totalTokens * uint256(_totalReputation), factor));
+    uint256 reward = rewardPayoutCycles[_payoutId].amount * sharePercentage / (10 ** (factor / 2));
```
Comment (area) [inline contracts/ColonyFunding.sol]: It also feels like I would be able to cook up an overflow here.

File: test/colony-funding.js [hunk 10]
```diff
@@ -403,4 +407,367 @@ contract("Colony Funding", () => {
...
+          .sqrt();
+        const reward = amount.mul(sharePercentage).round();
+
+        await newColony.claimRewardPayout(payoutId, data.userReputation.toString(), data.totalReputation.toString(), {
+          from: userAddress1
+        });
+
+        const rewardPayoutInfoAfterPayout = await newColony.getRewardPayoutInfo(payoutId);
+        const remaining = rewardPayoutInfoAfterPayout[2];
+
```
Comment (area) [inline test/colony-funding.js]: Can we print out how close we were explicitly for these tests?

File: contracts/ColonyFunding.sol (hunk 6 above)
Comment (filiplazovic) [reply contracts/ColonyFunding.sol]: Sorry, i should have checked the documentation.

File: contracts/ColonyFunding.sol (hunk 7 above)
Comment (filiplazovic) [reply contracts/ColonyFunding.sol]: I thought it was 77 (idiot me), and 1 will be added in `SafeMath.percent`. Which is not good, anyway.
About the underflows and overflows i was not so sure, thats why i didn't added the tests for more extreme values.
We can can continue this on Gitter, i just realized a few more things.

File: test/colony-funding.js (hunk 8 above)
Comment (filiplazovic) [reply test/colony-funding.js]: Sure, i just deleted them because i was getting `0` reward (As you mentioned above)

**Comment by filiplazovic** [comment]
@area Can you take a look at this?

File: test/colony-funding.js [hunk 11]
```diff
@@ -403,4 +406,290 @@ contract("Colony Funding", () => {
...
+      await checkErrorRevert(colony.startNextRewardPayout(otherToken.address), "colony-reward-payout-token-active");
+    });
+
+    it("should be able to collect rewards from multiple payouts of different token", async () => {
+      const tokenArgs = getTokenArgs();
+      const newToken = await Token.new(...tokenArgs);
+      await fundColonyWithTokens(colony, newToken, initialFunding.toString());
+
+      const payoutId1 = await colony.startNextRewardPayout.call(newToken.address);
+      await colony.startNextRewardPayout(newToken.address);
```
Comment (elenadimitrova) [inline test/colony-funding.js]: I'd argue to not return anything out of that function and instead get the `payoutId` via emitting an `event` out of the transaction itself. See https://github.com/JoinColony/colonyNetwork/blob/develop/test/colony-network-auction.js#L92

**Comment by area** [comment]
I like the choice of implementation 3 from the [repository we had on the side](https://github.com/area/payoutCalculationComparison); I think that implementation has the properties I'm happiest with. Two things about the tests are giving me pause though:

1. I like tests to be as deterministic as possible - no randomness! I also don't think we need this many; the important regimes to test are very small numbers (for accuracy) and very large ones (for overflow); I would expect to see tests using numbers like `1`, `10`, `1e76`, `1e77` and not much in between. 
2. The amount the solidity implementation is wrong by appears to be positive, which is definitely bad - if true, it'd mean that the last person to claim wouldn't be able to claim what was owed to them. That wasn't the case in the other repository, though, so something a little odd is going on here.

Finally, gas costs for this implementation are higher than I was hoping for. A way to reduce them is to have the user supply the values of the sqrt, which are then verified by squaring the passed claimed square root and checking that it is less than the original value. This calculation can be done automatically by our client, so it won't be more onerous for the user, and should reduce gas costs drastically.

**Comment by filiplazovic** [comment]
1. Im ok with that.
2. I think you are wrong, otherwise the tests would fail. I explicitly added `sqrt(denomerator) + 1` [here](https://github.com/JoinColony/colonyNetwork/pull/195/files#diff-8920cdf7ab56c438c907df9bd37d2931R193) in effort to avoid that. It seems like its doing the job, unless you know a case when it wont work.

Thanks for the hint about gas costs

File: contracts/IColony.sol [hunk 12]
```diff
@@ -275,7 +275,49 @@ contract IColony {
...
   /// @param _token Address of the token, `0x0` value indicates Ether
   function claimPayout(uint256 _id, uint256 _role, address _token) public;
-  
+
+  /// @notice Start next reward payout in `_token`. All funds in that are in reward pot at the time will become unavailable.
+  /// All tokens will be locked, and can be unlocked by calling `waiveRewardPayout` or `claimRewardPayout`.
+  /// @param _token Addess of the token used for reward payout
+  function startNextRewardPayout(address _token) public returns (uint256);
+
+  /// @notice Claim the reward payout at `_payoutId`. User needs to provide his reputation and colony-wide reputation
```
Comment (elenadimitrova) [inline contracts/IColony.sol]: Just some notes on comments language:
- Pls use the non-gender specific `their` instead of `his` or `hers`. There may be women interacting with this contract you know :)

File: contracts/ColonyStorage.sol [hunk 13]
```diff
@@ -44,6 +44,26 @@ contract ColonyStorage is DSAuth {
...
+    address tokenAddress;
+    // Unix timestamp of when the payout has started
+    uint256 blockTimestamp;
+  }
+
+  mapping (uint256 => RewardPayoutCycle) rewardPayoutCycles;
+  mapping (address => bool) activeRewardPayouts;
+  uint8 globalRewardPayoutCount;
+  mapping (address => uint256) userRewardPayoutCount;
+
```
Comment (elenadimitrova) [inline contracts/ColonyStorage.sol]: Comment mappings, arrays, global variables, everything really that is storage. As you see all other items, not just structs in `ColonyStorage` are commented.
Our current stand on documentation is that `IColony` and `IColonyNetwork` are the two interfaces we document plus the two respective storage contracts: `ColonyStorage` and `ColonyNetworkStorage` 
In addition any other non-EtherRouter contracts (not covered by the above) are documented e.g. `ReputationMiningCycle`

File: contracts/IColony.sol [hunk 14]
```diff
@@ -275,7 +275,49 @@ contract IColony {
...
+  /// @return Reputation root hash at the time of creation
+  /// @return Total colony tokens at the time of creation
+  /// @return Total amount of tokens taken aside for reward payout
+  /// @return Remaining (unclaimed) amount of tokens
+  /// @return Token address
+  /// @return Block number at the time of creation
+  function getRewardPayoutInfo(uint8 _payoutId) public view returns (bytes32, uint256, uint256, uint256, address, uint256);
+
+  /// @notice Moves remaining (unclaimed) tokens back to reward pot
+  /// Can only be called when reward payout cycle is finished i.e when 60 days passed from its creation
```
Comment (elenadimitrova) [inline contracts/IColony.sol]: when 60 days passed -> when 60 days _have_ passed

File: contracts/IColony.sol [hunk 15]
```diff
@@ -275,7 +275,49 @@ contract IColony {
...
+  /// @notice Claim the reward payout at `_payoutId`. User needs to provide his reputation and colony-wide reputation
+  /// which will be proven via Merkle proof inside this function.
+  /// Can only be called if payout is active, i.e when 60 days passed from its creation.
+  /// Can only be called if next in queue
+  /// @param _payoutId Id of the reward payout
+  /// @param _userReputation User reputation at the point of creation of reward payout cycle
+  /// @param _totalReputation Total reputation at the point of creation of reward payout cycle
+  function claimRewardPayout(uint8 _payoutId, int256 _userReputation, int256 _totalReputation) public;
+
+  /// @notice Waive reward payouts. This will unlock the senders tokens and increment users reward payout counter,
```
Comment (elenadimitrova) [inline contracts/IColony.sol]: senders tokens -> sender's tokens

File: contracts/IColony.sol [hunk 16]
```diff
@@ -275,7 +275,49 @@ contract IColony {
   /// @param _role Id of the role, as defined in `ColonyStorage` `MANAGER`, `EVALUATOR` and `WORKER` constants
   /// @param _token Address of the token, `0x0` value indicates Ether
   function claimPayout(uint256 _id, uint256 _role, address _token) public;
-  
+
+  /// @notice Start next reward payout in `_token`. All funds in that are in reward pot at the time will become unavailable.
```
Comment (elenadimitrova) [inline contracts/IColony.sol]: "in token" implies in some Token contract, pls replace with "for" or just say "Start next `_token` reward payout"

Comment (elenadimitrova) [reply contracts/IColony.sol]: Also here this reads weird
>All funds in that are in reward pot at the time 
Did you mean:
"All funds in the reward pot will become unavailable"

File: contracts/ColonyFunding.sol [hunk 17]
```diff
@@ -140,6 +141,94 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
+    //TODO: Prove that userReputation and totalReputation in reputationState are correct
+
+    uint256 userTokens = token.balanceOf(msg.sender);
+
+    require(_totalReputation > 0, "colony-reward-payout-invalid-total-reputation");
+    require(userTokens > 0, "colony-reward-payout-invalid-user-tokens");
+    require(_userReputation > 0, "colony-reward-payout-invalid-user-reputation");
+
+    uint256 numerator = mul(SafeMath.sqrt(uint256(_userReputation)), SafeMath.sqrt(userTokens));
+    uint256 denomerator = mul(SafeMath.sqrt(uint256(_totalReputation)), SafeMath.sqrt(payout.totalTokens));
```
Comment (elenadimitrova) [inline contracts/ColonyFunding.sol]: I don't know what `denomerator` is in English did you mean `denominator`?
